ingestion/ingest-google-docs.py

```diff
@@ -0,0 +1,156 @@
+
+def read_paragraph_element(element):
+    """Returns the text in the given ParagraphElement.
+
+        Args:
+            element: a ParagraphElement from a Google Doc.
+    """
+    text_run = element.get('textRun')
+    if not text_run:
+        return ''
+    return text_run.get('content')
+
+
+def read_structural_elements(elements):
+    """Recurses through a list of Structural Elements to read a document's text where text may be
+        in nested elements.
+
+        Args:
+            elements: a list of Structural Elements.
+    """
+    text = ''
+    for value in elements:
+        if 'paragraph' in value:
+            elements = value.get('paragraph').get('elements')
+            for elem in elements:
+                text += read_paragraph_element(elem)
+        elif 'table' in value:
+            # The text in table cells are in nested Structural Elements and tables may be
+            # nested.
+            table = value.get('table')
+            for row in table.get('tableRows'):
+                cells = row.get('tableCells')
+                for cell in cells:
+                    text += read_structural_elements(cell.get('content'))
+        elif 'tableOfContents' in value:
+            # The text in the TOC is also in a Structural Element.
+            toc = value.get('tableOfContents')
+            text += read_structural_elements(toc.get('content'))
+    return text
+
+def generate_secrets(token_pickle_path, raw_token_path, credentials_path, scope):
+    # Generate secrets to access Google API, if not already generated, otherwise load in 
+    creds = None
+    if os.path.exists(token_pickle_path):
+        with open(token_pickle_path, 'rb') as token:
+            creds = pickle.load(token)
+    # If there are no credentials available, let the user log in.
+    if not creds:
+        if creds and creds.expired and creds.refresh_token:
+            creds.refresh(Request())
+        else:
+            flow = client.flow_from_clientsecrets(credentials_path, scope)
+            store = file.Storage(raw_token_path)
+            creds = tools.run_flow(flow, store)
+        # Save the credentials for the next run
+        with open(token_pickle_path, 'wb') as token:
+            pickle.dump(creds, token)
+    if scope == SCOPE_READ_DRIVE:
+        service = build('drive', 'v3', credentials=creds)
+    elif scope == SCOPE_READ_DOCS:
+        service = build('docs', 'v1', credentials=creds)
+    else:
+        service = None
+    return service
+
+
+def run():
+    # Generate secrets, if not already generated
+    service_drive = generate_secrets(
+        'secrets/token_read_drive.pickle',
+        'secrets/token_read_drive.json',
+        'secrets/credentials.json',
+        SCOPE_READ_DRIVE
+        )
+    service_docs = generate_secrets(
+        'secrets/token_read_docs.pickle',
+        'secrets/token_read_docs.json',
+        'secrets/credentials.json',
+        SCOPE_READ_DOCS
+        )

```

Comment:
Thanks! This is a lot cleaner.

Reply:
For a future PR I'll put this and other constants in a YAML or something